src/state.rs

```diff
@@ -0,0 +1,199 @@
+
+use crate::{
+    custom::CustomEventContent,
+    error::{InvalidEvent, InvalidEventKind},
+    room::{
+        aliases::AliasesEventContent,
+        avatar::AvatarEventContent,
+        // canonical_alias::CanonicalAliasEventContent, create::CreateEventContent,
+        encryption::EncryptionEventContent,
+        guest_access::GuestAccessEventContent,
+        history_visibility::HistoryVisibilityEventContent,
+        join_rules::JoinRulesEventContent,
+        // member::MemberEventContent, name::NameEventContent,
+        // pinned_events::PinnedEventsEventContent, power_levels::PowerLevelsEventContent,
+        // server_acl::ServerAclEventContent, third_party_invite::ThirdPartyInviteEventContent,
+        tombstone::TombstoneEventContent,
+        topic::TopicEventContent,
+    },
+    EventContent, EventType, RoomEventContent, StateEventContent,
+};
+
+/// A state event.
+#[derive(Clone, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum AnyStateEvent {
+    /// m.room.aliases
+    RoomAliases(StateEvent<AliasesEventContent>),
+    // /// m.room.avatar
+    // RoomAvatar(StateEvent<AvatarEventContent>),
+
+    // /// m.room.canonical_alias
+    // RoomCanonicalAlias(StateEvent<CanonicalAliasEventContent>),
+
+    // /// m.room.create
+    // RoomCreate(StateEvent<CreateEventContent>),
+
+    // /// m.room.encryption
+    // RoomEncryption(StateEvent<EncryptionEventContent>),
+
+    // /// m.room.guest_access
+    // RoomGuestAccess(StateEvent<GuestAccessEventContent>),
+
+    // /// m.room.history_visibility
+    // RoomHistoryVisibility(StateEvent<HistoryVisibilityEventContent>),
+
+    // /// m.room.join_rules
+    // RoomJoinRules(StateEvent<JoinRulesEventContent>),
+
+    // /// m.room.member
+    // RoomMember(StateEvent<MemberEventContent>),
+
+    // /// m.room.name
+    // RoomName(StateEvent<NameEventContent>),
+
+    // /// m.room.pinned_events
+    // RoomPinnedEvents(StateEvent<PinnedEventsEventContent>),
+
+    // /// m.room.power_levels
+    // RoomPowerLevels(StateEvent<PinnedEventsEventContent>),
+
+    // /// m.room.server_acl
+    // RoomServerAcl(StateEvent<ServerAclEventContent>),
+
+    // /// m.room.third_party_invite
+    // RoomThirdPartyInvite(StateEvent<ThirdPartyInviteEventContent>),
+
+    // /// m.room.tombstone
+    // RoomTombstone(StateEvent<TombstoneEventContent>),
+
+    // /// m.room.topic
+    // RoomTopic(StateEvent<TopicEventContent>),
+
+    // /// Any state event that is not part of the specification.
+    // CustomState(StateEvent<CustomEventContent>),
+}
+
+/// To-device event.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct StateEvent<C: StateEventContent> {
+    // TODO flatten fails when deserializing for some reason
+    /// Data specific to the event type.
+    #[serde(flatten, with = "crate::event_content_type_serde")]
+    pub content: C,
+    /// The globally unique event identifier for the user who sent the event.
```

Comment:
Could you put empty lines between all fields with doc comments (and also between all methods)? Just for easier readability.

```suggestion
    pub content: C,

    /// The globally unique event identifier for the user who sent the event.
```